publisher-sdk/src/test/java/com/criteo/publisher/LiveCdbCallListenerTests.kt

```diff
@@ -0,0 +1,235 @@
+    whenever(cdbResponse.timeToNextCall).thenReturn(1_000)
+    whenever(bidManager.isBidSilent(freshCdbResponseSlot)).thenReturn(false);
+
+    liveCdbCallListener.onCdbResponse(cdbRequest, cdbResponse)
+
+    verify(bidManager).setTimeToNextCall(1_000)
+    verify(bidManager, never()).setCacheAdUnits(listOf(freshCdbResponseSlot))
+    verify(bidListener, never()).onNoBid()
+    verify(bidListener).onBidResponse(freshCdbResponseSlot)
+    verify(bidLifecycleListener).onCdbCallFinished(cdbRequest, cdbResponse)
+  }
+
+  @Test
+  fun onBidResponse_givenEmptyResponseServedWithinTimeBudget_ThenDontCacheAndDontPassTheResponseThrough() {
+    whenever(cdbResponse.slots).thenReturn(listOf())
+    whenever(cdbResponse.timeToNextCall).thenReturn(1_000)
+    whenever(bidManager.isBidSilent(freshCdbResponseSlot)).thenReturn(false);
+
+    liveCdbCallListener.onCdbResponse(cdbRequest, cdbResponse)
+
+    verify(bidManager).setTimeToNextCall(1_000)
+    verify(bidManager, never()).setCacheAdUnits(listOf(freshCdbResponseSlot))
+    verify(bidListener, never()).onBidResponse(any())
+    verify(bidListener).onNoBid()
+    verify(bidLifecycleListener).onCdbCallFinished(cdbRequest, cdbResponse)
+  }
+
+  @Test
+  fun onBidResponse_givenSilentBidServedWithinTimeBudget_ThenDontCacheAndDontPassTheResponseThrough() {
+    whenever(cdbResponse.slots).thenReturn(listOf(freshCdbResponseSlot))
+    whenever(cdbResponse.timeToNextCall).thenReturn(1_000)
+    whenever(bidManager.isBidSilent(freshCdbResponseSlot)).thenReturn(true);
+
+    liveCdbCallListener.onCdbResponse(cdbRequest, cdbResponse)
+
+    verify(bidManager).setTimeToNextCall(1_000)
+    verify(bidLifecycleListener).onCdbCallFinished(cdbRequest, cdbResponse)
+    verify(bidListener, never()).onBidResponse(any())
+    verify(bidManager, never()).setCacheAdUnits(listOf(freshCdbResponseSlot))
+    verify(bidListener).onNoBid()
+  }
+
+  @Test
+  fun onBidResponse_givenTimeBudgetExceededAndValidCacheEntry_ThenReturnCachedResponseAndUpdateCache() {
+    whenever(cdbResponse.slots).thenReturn(listOf(freshCdbResponseSlot))
+    whenever(cdbResponse.timeToNextCall).thenReturn(1_000)
+    whenever(bidManager.consumeCachedBid(cacheAdUnit)).thenReturn(cachedCdbResponseSlot)
+
+    liveCdbCallListener.onTimeBudgetExceeded()
+    liveCdbCallListener.onCdbResponse(cdbRequest, cdbResponse)
+
+    verify(bidManager).setTimeToNextCall(1_000)
+    verify(bidManager).consumeCachedBid(cacheAdUnit)
+    verify(bidManager).setCacheAdUnits(cdbResponse.slots)
+    verify(bidListener).onBidResponse(cachedCdbResponseSlot)
+    verify(bidListener, never()).onBidResponse(freshCdbResponseSlot)
+    verify(bidListener, never()).onNoBid()
+    verify(bidLifecycleListener).onCdbCallFinished(cdbRequest, cdbResponse)
+    verify(bidLifecycleListener).onBidConsumed(cacheAdUnit, cachedCdbResponseSlot)
+  }
+
+  @Test
+  fun onBidResponse_givenTimeBudgetExceededAndCacheEntryIsSilentBid_ThenDontReturnCachedEntryAndCacheNewResponse() {
```

Comment:
Yeah, if you are ok with this let's keep this as is and let's wait for EE-1276 where I am implementing the entire bid silence logic like what was drawn on the diagram (otherwise the PR will be even bigger).

Reply:
Ok, could you just add a TODO then ?